### ISSUE TITLE
explore split-ssh implementation

### ISSUE THREAD

**Issue by mfc** [QubesOS/qubes-issues#1962]
this feature would appeal to developers and sysadmins, the idea is to implement a feature for SSH similar to that which current exists for [GPG](https://www.qubes-os.org/doc/split-gpg/). 

This would protect users against SSH vulnerabilities including these recent ones:
* http://arstechnica.com/security/2015/07/bug-in-widely-used-openssh-opens-servers-to-password-cracking/
* http://arstechnica.co.uk/security/2016/01/bug-that-can-leak-crypto-keys-just-fixed-in-widely-used-openssh/

it has been referenced previously on the qubes mailing lists:
* https://groups.google.com/d/msg/qubes-users/ZYqMF0oqCP8/bJ5iRsdFAQAJ

**Comment by marmarek** [comment]
Should be pretty simple using ssh-agent and socat. Good material for community contribution.

**Comment by evadogstar** [comment]
This issue can be  solved by moving to New GPG version and not
need to realize something for SSH as suggested, because of GnuPG 2.1 seems support
them by default with --enable-ssh-support key
 
Proof link, search for `ssh` term on the page:
https://gnupg.org/faq/whats-new-in-2.1.html

**Comment by andrewdavidwong** [comment]
Would you be interested in doing this, @evadogstar?

**Comment by evadogstar** [comment]
Currently, I do not know the Qubes system as well as need to do that.

https://github.com/QubesOS/qubes-issues/issues/474
Related ticket! Link it please.

And @marmarek say that 

> While we use GPG 2.1 (as it is in Fedora 23), split-gpg doesn't take advantage of its new agent architecture. Take a look at linked PoC above.

But why I'm on 3.2RC1 on dom0 have old version? 
https://i.imgur.com/nsPuKFK.png

**Comment by marmarek** [comment]
On Tue, Jul 05, 2016 at 10:09:07AM -0700, evadogstar wrote:
> But why I'm on 3.2RC1 on dom0 have old version? 
> https://i.imgur.com/nsPuKFK.png

1. It's irrelevant what gpg version is in dom0 for split-gpg/split-ssh
2. Take a look at `gpg2` (it's how Fedora package it).

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab
A: Because it messes up the order in which people normally read text.
Q: Why is top-posting such a bad thing?

**Comment by evadogstar** [comment]
gpg-agent can work as ssh-agent after the feature enabled at `~/.gnupg/gpg-agent.conf` with `enable-ssh-support`.
To connect to gpg-agent ssh-client use this socket: SSH_AUTH_SOCK=$HOME/.gnupg/S.gpg-agent.ssh 

I suppose that 
1. Qubes gpg-split-client must open socket at AppVM and setup environment SSH_AUTH_SOCK with it for ssh-client
2. Then forward this socket to `gpg-agent` at gpg-domain(vault) through qubes-split-gpg-server 
3. All other key management carried at gpg-domain with standard gpg tools. Authorization with  OpenGPG (key/subkey) is also possible and it's very interesting feature instead of using only regular SSH keys.

Is it secure to forward ssh-client(appvm) socket to gpg-agent at vault? 
Transferred data must pass some validation before it will be transferred to gpg-agent from qubes-gpg-client?

**Comment by sjug** [comment]
Not sure why this is not a higher priority feature?

What is the proposed secure solution until this is fixed? Just have private ssh keys in an AppVm?

**Comment by andrewdavidwong** [comment]
@sjug: I've increased the priority, but bear in mind that this is a "help wanted" issue, which means that we'll require help from the community in order to implement it.

**Comment by jpouellet** [comment]
From the [OpenSSH 7.4 release notes](https://www.openssh.com/txt/release-7.4):
```
New Features
------------

 * ssh(1): Add a proxy multiplexing mode to ssh(1) inspired by the
   version in PuTTY by Simon Tatham. This allows a multiplexing
   client to communicate with the master process using a subset of
   the SSH packet and channels protocol over a Unix-domain socket,
   with the main process acting as a proxy that translates channel
   IDs, etc.  This allows multiplexing mode to run on systems that
   lack file- descriptor passing (used by current multiplexing
   code) and potentially, in conjunction with Unix-domain socket
   forwarding, with the client and multiplexing master process on
   different machines. Multiplexing proxy mode may be invoked using
   "ssh -O proxy ..."
```

**Comment by henn** [comment]
FYI - I put something together for this, though it's a ways from being as nicely packaged/documented as split-gpg. Pull requests welcomed!

You can find it at https://github.com/henn/qubes-app-split-ssh

**Comment by marmarek** [comment]
Thanks @henn!
I was wondering about some active proxy for ssh-agent - for example to filter requests based on public key requested, or to limit requests count in a single connection. But unfortunately [the protocol](http://api.libssh.org/rfc/PROTOCOL.agent) doesn't look friendly for this.

**Comment by henn** [comment]
Thanks, @marmarek!

To stop connection-reuse, I think either "ssh-add -c" or implementing a proxy could work, though for now I'm happy since VMs have to ask at least once. It looks like [this project](https://github.com/blueboxgroup/sshagentmux) has python code for a proxy that could be repurposed pretty easily.

RE: which key was requested, I picture the user having separate ssh-vaults for each private key, so haven't worried about this at this point.

Do you have any suggestions on how I could better package these scripts for distribution?

Also, I tried to automate the initial "ssh-add" for the ssh-vault by adding it to the `/rw/config/rc.local` file, but it didn't work (I think because of some combination of it being too early in the boot and not running within the GUI which set $SSH_AUTH_SOCK).
Do you have any suggestions on how to run something at AppVM startup under the GUI?

**Comment by mig5** [comment]
@henn I found I could auto-add the SSH key by creating a ~/.config/autostart/*.desktop file

```
[Desktop Entry]
Name=ssh-add
Exec=ssh-add
Type=Application
```

**Comment by jpouellet** [comment]
xref discussion of @henn's impl: https://groups.google.com/d/topic/qubes-users/TPZBZlixqsI/discussion

**Comment by jpouellet** [comment]
xref env var propagation: https://groups.google.com/d/topic/qubes-devel/gXj5KWSkPIM/discussion

**Comment by reynir** [comment]
Hi, I've been working on an implementation in OCaml using Mirage: https://github.com/reynir/qubes-mirage-ssh-agent

It's very much work in progress and there are some bugs, but it works well enough that I have been using it for a few days with two ssh keys. Building it from scratch without an existing OCaml setup is some work and the documentation could be improved in that regard. Feedback appreciated!

**Comment by pratheekrebala** [comment]
Are there any new developments on this task? I'm very much interested in this functionality and would like to help implement this.

**Comment by xaki23** [comment]
> Are there any new developments on this task? I'm very much interested in this functionality and would like to help implement this.

@reynir solution works like a charm.
i have been using it since late 2017/early 2018 "in production", changed from "deploy as a vm kernel" to "templated" (using builder-mirage supplied by @marmarek ) in late 2018.
it crashed on me like once in 2019 (requiring a restart of the agent vm. *gasp*), i reported that to reynir and he (probably correctly) pointed out it may have been fixed mid-2019, i was using a build build from very early 2019.
then some hilarity ensued because for some reason it now needs 24MB of ram for the agent VM (instead of 23MB before), and the error message from pvgrub is very high on my "most useless messages ever" list. (this only applies to templated installs, if you boot it as a kernel no pvgrub is involved)

in summary, if you want to help, helping with documentation would be most useful, since what seems to be missing is to make this more accessible for users.

**Comment by pratheekrebala** [comment]
Thanks for the update @xaki23! I'll go through @reynir's solution this week to better understand how it is implemented and work on improving documentation where possible.

**Comment by eduncan911** [comment]
I've used this as well for nearly two years (have since moved off of qubes for a bit, I'll be back when I'm done with other projects).

I think the main, work, is around integration with qubes core, like how `split-pgp` is a first-class citizen of Qubes.  It would ideal if this tooling was as well.

**Comment by mfc** [comment]
FYI here is walkthrough for setting up split-ssh manually: https://kushaldas.in/posts/using-split-ssh-in-qubesos-4-0.html

**Comment by fepitre** [comment]
Following my message on the list (https://github.com/fepitre/qubes-diagrams/tree/newversion) does anyone is interested in proposing a working updated implementation?

**Comment by arkenoi** [comment]
what about split-pkcs11?

**Comment by ben-grande** [comment]
Some alternatives that allow controlling which keyring each client has access via Qrexec arg:

- https://github.com/unman/qubes-ssh-agent
- https://github.com/ben-grande/qusal/tree/main/salt/sys-ssh-agent#access-control

But none of the projects posted on this thread solve the biggest problem, there is no filter proxy between the client and the agent, the client can issue any command to the agent:

- Lock the agent with `ssh-add -X`; and
- Delete keys from memory by issuing `ssh-add -D`

**Comment by marmarek** [comment]
The simple version got packaged at https://github.com/QubesOS/qubes-app-linux-split-ssh/pull/1
